Add support for bzlmod

jwnrt opened the pull request:
## Summary

Hi, this PR adds bzlmod support for this library and resolves #39.

I've had to update some dependencies to match the versions available in the bazel central registry. Where there was no semver-compatible version in the registry, I've updated the dependency to its latest version. The versions in the `WORKSPACE` and `MODULE.bazel` should match.

I've also updated some attribute names to use `allowlist` which extends this library's support to Bazel v7.0, but drops support for Bazel versions pre June 2020.

## Help needed

Buildifier built from source isn't available in the BCR right now, but a prebuilt binary is. I hope this change is acceptable, but at the same time it breaks the example and I'm not sure how to resolve the error.

The binary has a [default_runfiles](https://github.com/google/bazel_rules_install/blob/e12fa95adda9fae107a22798acc2a3849d46f94a/installer/def.bzl#L32) file at `external/buildifier_prebuilt~6.4.0~buildifier_prebuilt_deps_extension~buildifier_linux_amd64/file/buildifier`, but the [`rlocation` call here](https://github.com/google/bazel_rules_install/blob/e12fa95adda9fae107a22798acc2a3849d46f94a/installer/installer.bash.template#L79) for this path returns empty.

Should this file just be skipped if it has no `rlocation`? I'm no Bazel expert so admit I don't know exactly what's going on here.

Thanks!

jwnrt committed:
Update stardoc to v0.6.2

This version is in the Bazel Central Registry.

jwnrt committed:
Adopt allowlist attr names

Bazel moved to using "allowlist" in Jun 2020 and dropped support for the
legacy names in Sep 2023, so newer versions of Bazel don't work.

jwnrt committed:
Add support for bzlmod modules

bttk commented on MODULE.bazel:
```diff
@@ -0,0 +1,52 @@
+# Copyright 2018 The Bazel Authors.
```
Here and in other new files use the year this file is created (2024) instead of 2018

bttk commented on WORKSPACE:
```diff
@@ -32,11 +32,11 @@ local_repository(
-    name = "com_github_bazelbuild_buildtools",
-    sha256 = "e3bb0dc8b0274ea1aca75f1f8c0c835adbe589708ea89bf698069d0790701ea3",
-    strip_prefix = "buildtools-5.1.0",
+    name = "buildifier_prebuilt",
```
I'd like to approve this change, but I don't see many Google projects using this binary dependency.

@hzeller What do you think?

hzeller replied:
We can apt-get a buildifier on the CI if the script/run_buildifier.sh is to be run there.

jwnrt replied:
I've updated the PR to install buildifier with `go install` (I couldn't find it in the apt repositories)

bttk reviewed (changes requested):
Thank you for this PR!

I'm not sure about the introduction of buildifier_prebuilt
Without it, looks good!

hzeller reviewed (commented):
Nice! Sorry for the delay, github notifications are easy to miss, so I stumbled upon this on accident.

It looks like it makes sense to review some of the dependencies and prune where possible. Less things we depend on the less things can break in the future in weird ways :)

jwnrt committed:
Remove buildifier dependency

hzeller commented on WORKSPACE:
```diff
@@ -32,29 +32,29 @@ local_repository(
     ],
 )
 
 http_archive(
```
That sounds like a promising direction

jwnrt committed:
Replace buildtools with hello_world in example

jwnrt committed:
Run system buildifier in CI

jwnrt commented:
I've updated the PR to remove `buildifier` entirely. Development (and CI) now uses the system buildifier for formatting. I've changed the example which installed buildifier to now install a hello-world C binary.

hzeller commented:
Less dependencies is good, yay. What do you think @bttk ?

I wonder why the bazelmod.lock files are so gigantic. It makes it essentially impossible to review and look out for potential security risks of whatever it imports. Can these be trimmed to the minimum needed ? (I don't know anything about bzlmod).

jwnrt commented:
The lock files are indeed huge, but I don't see any options for compressing or shortening them. It looks like they include all of the built-in dependencies like `bazel_tools` dependencies which probably contributes to their size quite a lot.

See https://github.com/bazelbuild/bazel/issues/19971 for an open issue about removing built-ins from the lock.

It looks like `rules_python` has [chosen to exclude the lock entirely](https://github.com/bazelbuild/rules_python/pull/1680). What do you think about that approach? I think we would add `--lockfile_mode=off` to a `.bazelrc` file in this repo, and add the locks to `.gitignore` just in case.

hzeller commented:
I think I like that approach not to include the *.lock file and consider it an ephemeral artifact.
(In particular since this particular project is not super-sensitive to super-exact versions of what we depend on. Essentially, as long as a shell script can be executed, we're good).

hzeller commented:
Nice!

I think it would be good if we mention the range of versions supported of bazel with this.
This repo should just work from bazel 5 to 7 right ?

Anything you'd like to add @bttk ? Are we ready to merge ?

bttk reviewed (approved):
LGTM!

hzeller commented:
Thanks for the PR, merged!
